components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -2053,19 +2062,92 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
+            // dependencies (DB) fail.
+            checkAgainstDB(),
+
+            // Centralized perms checks only throw, when an interaction error occurs - connection not available or similar.
+            // When the user is not permitted to perform the action, the call will resolve, encoding the result in the response.
+            checkWithCentralizedPerms(),
+        ]);
+
+        // check against DB resolved, which means the user is permitted to perform the action
+        if (fromDB.status === "fulfilled") {
+            if (fromCentralizedPerms.status === "fulfilled") {
+                // we got a result from centralized perms, but we still need to check if the outcome was such that the user is permitted
+                reportCentralizedPermsValidation(fineGrainedOp, fromCentralizedPerms.value.permitted === true);
+            } else {
+                // centralized perms promise rejected, we do not have an agreement
+                reportCentralizedPermsValidation(fineGrainedOp, false);
+            }
+
+            // Always return the result from the DB check
+            return fromDB.value;
+        } else {
+            // The check agains the DB failed. This means the user does not have access.
+
+            if (fromCentralizedPerms.status === "fulfilled") {
+                // we got a result from centralized perms, but we still need to check if the outcome was such that the user is NOT permitted
+                reportCentralizedPermsValidation(fineGrainedOp, fromCentralizedPerms.value.permitted === false);
+            } else {
+                // centralized perms promise rejected, we do not have an agreement
+                reportCentralizedPermsValidation(fineGrainedOp, false);
+            }
+
+            // We re-throw the error from the DB permission check, to propagate it upstream.
+            throw fromDB.reason;
         }
+    }
+
+    protected async guardOrganizationOperationWithCentralizedPerms(
+        orgId: string,
+        op: OrganizationOperation,
+    ): Promise<CheckResult> {
+        const user = this.checkUser();
 
-        const members = await this.teamDB.findMembersByTeam(team.id);
-        await this.guardAccess({ kind: "team", subject: team, members }, op);
-        return { team, members };
+        switch (op) {
+            case "org_metadata_read":
+                return await this.authorizer.check(ReadOrganizationMetadata(user.id, orgId));
+            case "org_metadata_write":
+                return await this.authorizer.check(WriteOrganizationMetadata(user.id, orgId));
+
+            case "org_members_read":
+                return await this.authorizer.check(ReadOrganizationMembers(user.id, orgId));
+            case "org_members_write":
+                return await this.authorizer.check(WriteOrganizationMembers(user.id, orgId));
+
+            default:
+                return NotPermitted;
+        }
```

Comment:
This acts as an adapter between the "fine grained ops" defined as strings, and the actual checks. Once we have fully migrated, this adapter won't exist, but it makes the existing code a bit easier to evolve, and reason about.

Reply:
Are we going to pass the check constants instead fo the strings? Would be great to move this into the authorization folder as well.

Reply:
Eventually will go away, once we migrate but for now, they are useful as a way of adapting the behaviour.